governance/program/src/state/enums.rs

```diff
@@ -92,3 +92,60 @@ impl Default for ProposalState {
         ProposalState::Draft
     }
 }
+
+/// The type of the vote threshold percentage used to resolve a vote on a Proposal
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq, BorshDeserialize, BorshSerialize, BorshSchema)]
+pub enum VoteThresholdPercentageType {
+    /// Voting threshold of Yes votes in % required to tip the vote
+    /// It's the percentage of tokens out of the entire pool of governance tokens eligible to vote
+    /// Note: If the threshold is below or equal to 50% then an even split of votes ex: 50:50 or 40:40 is always resolved as Defeated
+    /// In other words +1 vote tie breaker is always required to have a successful vote
+    YesVote,
+
+    /// The minimum number of votes in % out of the entire pool of governance tokens eligible to vote
+    /// which must be cast for the vote to be valid
+    /// Once the quorum is achieved a simple majority (50%+1) of Yes votes is required for the vote to succeed
+    /// Note: Quorum is not implemented in the current version
+    Quorum,
+}
+
+/// The source of voter weights used to vote on proposals
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq, BorshDeserialize, BorshSerialize, BorshSchema)]
+pub enum VoteWeightSource {
+    /// Governing token deposits into the Realm are used as voter weights
+    Deposit,
+    /// Governing token account snapshots as of the time a proposal entered voting state are used as voter weights
+    /// Note: Snapshot source is not supported in the current version
+    /// Support for account snapshots are required in solana and/or arweave as a prerequisite
+    Snapshot,
```

Comment:
How would the design need to change to properly support this? Would you need to change any other structs to properly snapshot the accounts at the time the proposal was created?

Reply:
This is a difficult one because we are talking about a feature which doesn't exist yet. However my take on it is that it should be externalised and `Governance` program would make a CPI call to get voter's vote weight.
For example: 
1) We can create `Snapshot` program which would create and store account snapshot natively on Solana. 
2) We store snapshots on arweave and make calls to a program that would allow us to read the snapshots from there.

In the cases above all we need is `voting_at_slot` field so we know the `as_of_slot` we should query the snapshots for and that field was added in this PR. Other than that I think we don't need any struct changes.

Saying all that, I'm not sure about this snapshot strategy. This is how it's done on `ETH` but it allows you to play the system and create a proposal at a time convenient for you. 
Some other ideas from users are for example to use the time you have staked your tokens in a pool as a weight on top of the amount of tokens. But in this case I also think it'll be just a CPI call to whatever program can determine the weight.